src/Npgsql/NpgsqlParameter.cs

```diff
@@ -368,7 +368,22 @@ public NpgsqlDbType NpgsqlDbType
             {
                 if (_dataTypeName != null)
                     return _dataTypeName;
-                throw new NotImplementedException("Infer from others");
+                if (_npgsqlDbType.HasValue)
+                {
+                    var primitiveNpgsqlType = _npgsqlDbType.Value & ~NpgsqlDbType.Array & ~NpgsqlDbType.Range;
+                    var primitiveTypeName = GlobalTypeMapper.Instance.ToPgTypeName(primitiveNpgsqlType);
+
+                    if ((_npgsqlDbType.Value & NpgsqlDbType.Array) != 0 ||
+                        (_npgsqlDbType.Value & NpgsqlDbType.Range) != 0)
+                    {
+                        primitiveTypeName += "[]";
+                    }
+
+                    return primitiveTypeName;
+                }
+                if (_value != null)   // Infer from value
```

Comment:
On more point: a complete implementation would also infer DataTypeName from DbType, not only NpgsqlDbType.

Reply:
Inference from DbType included in 457d0e3